pkg/controller/schedulingmanager/controller.go

```diff
@@ -93,53 +110,73 @@ func newController(config *util.ControllerConfig) (*SchedulerController, error)
 	return c, nil
 }
 
+func (c *SchedulingManager) GetScheduler(schedulingKind string) *SchedulerWrapper {
+	c.RLock()
+	defer c.RUnlock()
+	scheduler, ok := c.schedulers[schedulingKind]
+	if !ok {
+		return nil
+	}
+	return scheduler
+}
+
 // Run runs the Controller.
-func (c *SchedulerController) Run(stopChan <-chan struct{}) {
+func (c *SchedulingManager) Run(stopChan <-chan struct{}) {
 	go c.controller.Run(stopChan)
 
 	// wait for the caches to synchronize before starting the worker
 	if !cache.WaitForCacheSync(stopChan, c.controller.HasSynced) {
-		runtime.HandleError(errors.New("Timed out waiting for cache to sync"))
+		runtime.HandleError(errors.New("Timed out waiting for cache to sync in scheduling manager"))
 		return
 	}
 
 	c.worker.Run(stopChan)
+
+	go func() {
+		<-stopChan
+		c.shutdown()
+	}()
+}
+
+func (c *SchedulingManager) shutdown() {
+	for _, scheduler := range c.schedulers {
+		// Indicate scheduler and associated goroutines to be stopped in schedulingpreference controller.
+		close(scheduler.stopChan)
```

Comment:
we need to stop plugins if shutdown.

Reply:
That was the leak in earlier code :-), which is now stopped via this stopchannel [here](https://github.com/kubernetes-sigs/federation-v2/blob/1804943a6eed0d349f0f930c39f439c094e71378/pkg/controller/schedulingpreference/controller.go#L175) and [here](https://github.com/kubernetes-sigs/federation-v2/blob/1804943a6eed0d349f0f930c39f439c094e71378/pkg/schedulingtypes/replicascheduler.go#L149).

Reply:
got it. Thanks.